R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+    dep})
+  dependencies <- lapply(dependencies, replace_with_lock, lock)
+  list(package = package, dependencies = dependencies)
+}
+
+strip_available_dependencies <- function(dependencies) {
+  dependencies <- lapply(dependencies
+    , function(package) {
+      if(package$name %in% pesky_namespaces) NULL
+      else package
+    })
+ dependencies[!vapply(dependencies, is.null, logical(1))]
+}
+
+#' We don't trust package authors to only put a package in once
+strip_duplicate_dependencies <- function(dependencies) {
+  dependencies[!duplicated(vapply(dependencies, function(d) d$name
+    , character(1)))]
+}
+
+#' Get the dependencies for a given package
+get_remote_dependencies <- function(package) {
+  UseMethod("get_remote_dependencies")
+}
+
+#' If a package is local we just read from the directory given
+get_remote_dependencies.local <- function(package) {
+  description_name <- paste0(package$dir, .Platform$file.sep, "DESCRIPTION")
+  list(package = package
+    , dependencies =
+      dependencies_from_description(package, read.dcf(description_name)))
+}
+
+#' For packages on CRAN we will extract to a temporary directory when we
+#' download the accurate remote DESCRIPTION file. Because these are tarballs
+#' there is no simple way to extract only our desired file like we can with
+#' zipfiles using the unz function.
+get_remote_dependencies.CRAN <- function(package) {
+  original_version <- package$version
+  if (package$is_dependency_package) {
+    package$version <- NA
+  }
+  filepath <- download_package(package)
+
+  sep <- .Platform$file.sep
+  split_fp <- strsplit(filepath,sep)[[1]]
+  dirpath <- dirname(filepath)
+  file_list <- untar(filepath, list = TRUE)
+  description_name <- file_list[grepl(paste0("^[^", sep, "]+", sep
+    ,"DESCRIPTION$"), file_list)]
+  output <- untar(filepath, description_name, exdir = dirpath)
+  description_path <- paste0(dirpath, sep, description_name)
+  package$download_path <- filepath
+  package$version <- original_version
+  dcf <- read.dcf(file = description_path)
+  package$latest_version <- version_from_description(package$name, dcf)
+  unlink(description_path)
+  list(package = package, dependencies = dependencies_from_description(package, dcf))
+}
+
+#' Download the accurate remote DESCRIPTION file for a github repo.
+get_remote_dependencies.github <- function(package) {
+  output <- download_description_github(package)
+  list(package = output$package
+    , dependencies = dependencies_from_description(package, output$dcf))
+}
+
+download_description_github <- function(package) {
+  remote <- package$remote
+  filepath <- download_package(structure(
+    package,
+    class = c(remote, class(package))))
+  package$download_path <- filepath
+  file_list <- unzip(filepath, list = TRUE)
+  sep <- .Platform$file.sep
+  subdir <- ""
+  if (!is.null(package$subdir)){
+    subdir <- paste0(sep, package$subdir)
+  }
+  description_name <- file_list$Name[grepl(paste0("^[^", sep, "]+"
```

Comment:
So these packages always unzip into some directory defined in the zip file, So I'm looking for non-file-separators followed by a separator ([^/]+/) to get the description file in the immediate directory above, unless there is the subdirectory, in which case it is two directories up.  It's almost guaranteed to be the first thing that matches DESCRIPTION at all in file_list$Name, but after seeing tons of weird cran and package behaviors I'm super paranoid and wanted to be very explicit.